### DISCUSSION TITLE
Mock fetch inside API route

### DISCUSSION THREAD

**Discussion by adrianrios25** [Xunnamius/next-test-api-route-handler#1125] [Q&A] [answered]
<details open><summary><strong>The problem</strong></summary>

I'm calling fetch api inside my api route so in my test I have tried something like this

```
it("test POST", async () => {
  global.fetch = jest.fn(() =>
	  Promise.resolve({
		  json: () =>
			  Promise.resolve({
				  authenticationStatus: "NO",
			  }),
	  })
  ) as jest.Mock;
  
  await testApiHandler({
	  appHandler,
	  requestPatcher(request) {
		  request.headers.set("CORRELATION-ID", "trx-123");
	  },
  
	async test({ fetch }) {
		const res = await fetch({
			method: "POST",
			body: JSON.stringify({
				key: "value",
			}),
		});
		await expect(res.json()).resolves.toStrictEqual({
			message: "Not Authorized",
		});
	},
  });
});
```
this is working when I run the test in local but in azure pipeline I'm getting this error. Is there a way to do this properly?

```
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From src/__test__/api/***/route.test.ts.

      at BufferListStream.Readable (node_modules/readable-stream/lib/internal/streams/readable.js:207:36)
      at BufferListStream.Duplex (node_modules/readable-stream/lib/internal/streams/duplex.js:50:12)
      at new BufferListStream (node_modules/bl/bl.js:33:16)
      at new IncomingMessageStream (node_modules/tedious/src/incoming-message-stream.ts:24:15)
      at new MessageIO (node_modules/tedious/src/message-io.ts:40:34)
      at Connection.socketHandlingForSendPreLogin (node_modules/tedious/src/connection.ts:2025:22)
      at node_modules/tedious/src/connection.ts:2115:12
```
</details>

</details>

**Comment by Xunnamius** [comment, marked as answer]
That's a strange error, and it doesn't seem like it has to do with NTARH directly if things are working in your local environment. Without a minimal example repo demonstrating this error, I can only make some guesses:

- Your test is trying to import something AFTER Jest has ended. This is a sign that you are not awaiting a promise somewhere where you should be. If this unawaited promise is in your test file or in your source files somewhere, I am not sure. I use eslint's promise rules to catch these types of errors since they can get pretty tough to debug manually, especially in a CI pipeline. There's also [`--detectOpenHandles`](https://jestjs.io/docs/cli#--detectopenhandles).

- Your simple `fetch` mock is hiding some sort of complexity when run locally that your software crashes into when run in your pipeline. Maybe.

- There is actually something wrong with NTARH, which would be interesting indeed!

**Reply by adrianrios25** [reply]
ok I have this figured out, and it's a coding error on my end, I am calling an async method and forgot to put the await keyword.
